### ISSUE TITLE
Workflow Enhancements

### ISSUE THREAD

**Issue by qqmyers** [IQSS/dataverse#5044]
As part of #4706, I needed the workflow mechanism to allow read-access to some settings and to enable internal workflows to act on behalf of a user (i.e. via access to their apiToken). In addition, I ran into some issue in which post-publication workflows only had access to the state of a dataset that did not include the publication updates (e.g. a change of the latest version to 'released') and, when resolving that, to some inconsistencies in adding and removing locks.

**Comment by pdurbin** [comment]
@qqmyers I moved this over to code review based on what you wrote at https://github.com/IQSS/dataverse/pull/5048#issuecomment-426732955 but if you don't mind I'm curious about what sort of workflows you're imagining. "As a curator, I want to..."

**Comment by qqmyers** [comment]
@pdurbin - this is part of the DPN submission of bagit bags #4706  - I needed to update the workflow mechanism itself to be able to get additional settings, for a post-publication workflow to be able to access the state of the dataset being published and further update it, etc.  As with the sha256/512 PR before, the changes here are used for/were motivated by DPN submission but are stand-alone technically.

**Comment by pdurbin** [comment]
@qqmyers perfect. And #4706 has tons of detail, especially the wiki page your wrote, on what DPN submission is about. Thanks!

**Comment by kcondon** [comment]
@qqmyers Do you mind updating this branch from /develop? I can't build/deploy due to old version. I've just verified rsync workflow/engine with existing code and am ready to verify/merge this branch. Thanks!

**Comment by kcondon** [comment]
@qqmyers Thanks!

**Comment by qqmyers** [comment]
@kcondon - After an error in which eclipse thought it merged with develop but left 100+ files out of synch, I've gone ahead and added a couple 'repair' commits to 'recover'. At this point, it should be merged except that it appears a few files have permission differences (no difference in source).  If you can fix those, this can go ahead. If you prefer, I can create a new branch.

**Comment by pdurbin** [comment]
@qqmyers I just took at look at pull request #5048 after @kcondon mentioned your comment above to me. It does seem to have weird "chmod" stuff going on, so we'd lose the ability to execute scripts. If it's not too much trouble, a fresh pull request would be appreciated!

**Comment by qqmyers** [comment]
https://github.com/IQSS/dataverse/pull/5224 is the cleaned up PR

**Comment by kcondon** [comment]
@qqmyers I am seeing some exceptions when attempting to use the rsync publish workflow that did not occur in /develop. Also, it fails to publish. I'll discuss  more with @pameyer to see if we can isolate further. Meanwhile here is the stack trace when I click publish:

Local Exception Stack: 
Exception [EclipseLink-5006] (Eclipse Persistence Services - 2.5.2.v20140319-9ad6abd): org.eclipse.persistence.exceptions.OptimisticLockException
Exception Description: The object [[DatasetVersion id:30]] cannot be updated because it has changed or been deleted since it was last read. 
Class> edu.harvard.iq.dataverse.DatasetVersion Primary Key> 30
	at org.eclipse.persistence.exceptions.OptimisticLockException.objectChangedSinceLastReadWhenUpdating(OptimisticLockException.java:144)
	at org.eclipse.persistence.descriptors.VersionLockingPolicy.validateUpdate(VersionLockingPolicy.java:790)
	at org.eclipse.persistence.internal.queries.DatabaseQueryMechanism.updateObjectForWriteWithChangeSet(DatabaseQueryMechanism.java:1087)
	at org.eclipse.persistence.queries.UpdateObjectQuery.executeCommitWithChangeSet(UpdateObjectQuery.java:84)
	at org.eclipse.persistence.internal.queries.DatabaseQueryMechanism.executeWriteWithChangeSet(DatabaseQueryMechanism.java:301)
	at org.eclipse.persistence.queries.WriteObjectQuery.executeDatabaseQuery(WriteObjectQuery.java:58)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:899)
	at org.eclipse.persistence.queries.DatabaseQuery.executeInUnitOfWork(DatabaseQuery.java:798)
	at org.eclipse.persistence.queries.ObjectLevelModifyQuery.executeInUnitOfWorkObjectLevelModifyQuery(ObjectLevelModifyQuery.java:108)
	at org.eclipse.persistence.queries.ObjectLevelModifyQuery.executeInUnitOfWork(ObjectLevelModifyQuery.java:85)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:2896)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1804)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1786)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1737)
	at org.eclipse.persistence.internal.sessions.CommitManager.commitChangedObjectsForClassWithChangeSet(CommitManager.java:267)
	at org.eclipse.persistence.internal.sessions.CommitManager.commitAllObjectsForClassWithChangeSet(CommitManager.java:192)
	at org.eclipse.persistence.internal.sessions.CommitManager.commitAllObjectsWithChangeSet(CommitManager.java:138)
	at org.eclipse.persistence.internal.sessions.AbstractSession.writeAllObjectsWithChangeSet(AbstractSession.java:4207)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.commitToDatabase(UnitOfWorkImpl.java:1441)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.commitToDatabaseWithPreBuiltChangeSet(UnitOfWorkImpl.java:1587)
	at org.eclipse.persistence.internal.sessions.RepeatableWriteUnitOfWork.writeChanges(RepeatableWriteUnitOfWork.java:452)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.flush(EntityManagerImpl.java:863)
	at com.sun.enterprise.container.common.impl.EntityManagerWrapper.flush(EntityManagerWrapper.java:437)
	at edu.harvard.iq.dataverse.workflow.WorkflowServiceBean.lockDataset(WorkflowServiceBean.java:288)
	at edu.harvard.iq.dataverse.workflow.WorkflowServiceBean.start(WorkflowServiceBean.java:101)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:73)
	at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at sun.reflect.GeneratedMethodAccessor128.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
	at sun.reflect.GeneratedMethodAccessor131.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
	at com.sun.ejb.containers.EjbAsyncTask.call(EjbAsyncTask.java:101)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
]]

[2018-10-22T18:41:50.876-0400] [glassfish 4.1] [WARNING] [AS-EJB-00056] [javax.enterprise.ejb.container] [tid: _ThreadID=146 _ThreadName=__ejb-thread-pool1] [timeMillis: 1540248110876] [levelValue: 900] [[
  A system exception occurred during an invocation on EJB WorkflowServiceBean, method: public void edu.harvard.iq.dataverse.workflow.WorkflowServiceBean.start(edu.harvard.iq.dataverse.workflow.Workflow,edu.harvard.iq.dataverse.workflow.WorkflowContext) throws edu.harvard.iq.dataverse.engine.command.exception.CommandException]]

[2018-10-22T18:41:50.876-0400] [glassfish 4.1] [WARNING] [] [javax.enterprise.ejb.container] [tid: _ThreadID=146 _ThreadName=__ejb-thread-pool1] [timeMillis: 1540248110876] [levelValue: 900] [[
  
javax.ejb.EJBException
	at com.sun.ejb.containers.EJBContainerTransactionManager.processSystemException(EJBContainerTransactionManager.java:748)
	at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:698)
	at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:503)
	at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4566)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2074)
	at com.sun.ejb.containers.EjbAsyncTask.call(EjbAsyncTask.java:114)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.persistence.OptimisticLockException: Exception [EclipseLink-5006] (Eclipse Persistence Services - 2.5.2.v20140319-9ad6abd): org.eclipse.persistence.exceptions.OptimisticLockException
Exception Description: The object [[DatasetVersion id:30]] cannot be updated because it has changed or been deleted since it was last read. 
Class> edu.harvard.iq.dataverse.DatasetVersion Primary Key> 30
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.flush(EntityManagerImpl.java:865)
	at com.sun.enterprise.container.common.impl.EntityManagerWrapper.flush(EntityManagerWrapper.java:437)
	at edu.harvard.iq.dataverse.workflow.WorkflowServiceBean.lockDataset(WorkflowServiceBean.java:288)
	at edu.harvard.iq.dataverse.workflow.WorkflowServiceBean.start(WorkflowServiceBean.java:101)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:73)
	at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at sun.reflect.GeneratedMethodAccessor128.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
	at sun.reflect.GeneratedMethodAccessor131.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
	at com.sun.ejb.containers.EjbAsyncTask.call(EjbAsyncTask.java:101)
	... 4 more
Caused by: Exception [EclipseLink-5006] (Eclipse Persistence Services - 2.5.2.v20140319-9ad6abd): org.eclipse.persistence.exceptions.OptimisticLockException
Exception Description: The object [[DatasetVersion id:30]] cannot be updated because it has changed or been deleted since it was last read. 
Class> edu.harvard.iq.dataverse.DatasetVersion Primary Key> 30
	at org.eclipse.persistence.exceptions.OptimisticLockException.objectChangedSinceLastReadWhenUpdating(OptimisticLockException.java:144)
	at org.eclipse.persistence.descriptors.VersionLockingPolicy.validateUpdate(VersionLockingPolicy.java:790)
	at org.eclipse.persistence.internal.queries.DatabaseQueryMechanism.updateObjectForWriteWithChangeSet(DatabaseQueryMechanism.java:1087)
	at org.eclipse.persistence.queries.UpdateObjectQuery.executeCommitWithChangeSet(UpdateObjectQuery.java:84)
	at org.eclipse.persistence.internal.queries.DatabaseQueryMechanism.executeWriteWithChangeSet(DatabaseQueryMechanism.java:301)
	at org.eclipse.persistence.queries.WriteObjectQuery.executeDatabaseQuery(WriteObjectQuery.java:58)
	at org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:899)
	at org.eclipse.persistence.queries.DatabaseQuery.executeInUnitOfWork(DatabaseQuery.java:798)
	at org.eclipse.persistence.queries.ObjectLevelModifyQuery.executeInUnitOfWorkObjectLevelModifyQuery(ObjectLevelModifyQuery.java:108)
	at org.eclipse.persistence.queries.ObjectLevelModifyQuery.executeInUnitOfWork(ObjectLevelModifyQuery.java:85)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:2896)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1804)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1786)
	at org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1737)
	at org.eclipse.persistence.internal.sessions.CommitManager.commitChangedObjectsForClassWithChangeSet(CommitManager.java:267)
	at org.eclipse.persistence.internal.sessions.CommitManager.commitAllObjectsForClassWithChangeSet(CommitManager.java:192)
	at org.eclipse.persistence.internal.sessions.CommitManager.commitAllObjectsWithChangeSet(CommitManager.java:138)
	at org.eclipse.persistence.internal.sessions.AbstractSession.writeAllObjectsWithChangeSet(AbstractSession.java:4207)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.commitToDatabase(UnitOfWorkImpl.java:1441)
	at org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.commitToDatabaseWithPreBuiltChangeSet(UnitOfWorkImpl.java:1587)
	at org.eclipse.persistence.internal.sessions.RepeatableWriteUnitOfWork.writeChanges(RepeatableWriteUnitOfWork.java:452)
	at org.eclipse.persistence.internal.jpa.EntityManagerImpl.flush(EntityManagerImpl.java:863)
	... 36 more
]]

**Comment by kcondon** [comment]
@qqmyers @pameyer an update on the above:
The optimistic lock error occurs on this branch when I click publish on a dataset with an rsync package file, using the rsync workflow. It does not occur in the same circumstance using /develop. These errors are usually timing issues it seems to me. Please note that no request appears to be sent to Rsal when that error happens. If you click publish again, not optimistic lock error and publish does work.

**Comment by qqmyers** [comment]
@kcondon - Where is the rsync workflow?

**Comment by kcondon** [comment]
@qqmyers https://github.com/IQSS/dataverse/blob/develop/scripts/api/data/workflows/internal-httpSR-workflow.json

**Comment by qqmyers** [comment]
@kcondon - Can you try with the new commits (note that there's an additional line in the update sql script to run. They include a possible fix as well as an update to pass the datasetExternallyReleased flag through to the resume phase of external workflows and removal of an unlock command in a rollback that I missed previously. 

I wasn't able to reproduce the error above, although I haven't fully duplicated the conditions. The error you show above is happening when the workflow starts, when it tries to create a lock, which happens before any workflow steps run. I set up the internal pause workflow as the default PrePublication workflow and didn't see an error. (I was also able to manually call the API and complete the workflow.) If it is a timing issue, the flush() I added before the workflow may be a fix, but I'm not sure. If you still see a problem, I can look further into it.

**Comment by kcondon** [comment]
@qqmyers you have fixed the optimistic lock problem but there is now an issue where the dataset lock for publish is not being set, allowing publish button to be pressed again, no feedback (publish lock banner not set)

Also, have noticed another issue that is actually in /develop already so may break out into a separate issue: when either uploading a package file or publishing a dataset with package file, see an exception about full-text indexing failing. We think it is because the package "file" is actually a directory and so cannot open it as you would a normal file:

[2018-10-23T16:42:31.583-0400] [glassfish 4.1] [INFO] [] [job-101] [tid: _ThreadID=136 _ThreadName=concurrent/__defaultManagedExecutorService-managedThreadFactory-Thread-1] [timeMillis: 1540327351583] [levelValue: 800] [[
  Successfully created a file of type package]]

[2018-10-23T16:42:32.546-0400] [glassfish 4.1] [WARNING] [] [edu.harvard.iq.dataverse.search.IndexServiceBean] [tid: _ThreadID=136 _ThreadName=concurrent/__defaultManagedExecutorService-managedThreadFactory-Thread-1] [timeMillis: 1540327352546] [levelValue: 900] [[
  Full-text indexing for X8KTAB failed]]

[2018-10-23T16:42:32.548-0400] [glassfish 4.1] [SEVERE] [] [] [tid: _ThreadID=136 _ThreadName=Thread-9] [timeMillis: 1540327352548] [levelValue: 1000] [[
  java.io.IOException: Failed to open local file file:///usr/local/glassfish4/glassfish/domains/domain1/files/10.5072/FK2/X8KTAB/166a2aa24bb-2c80fb40c79e
	at edu.harvard.iq.dataverse.dataaccess.FileAccessIO.open(FileAccessIO.java:100)
	at edu.harvard.iq.dataverse.search.IndexServiceBean.addOrUpdateDataset(IndexServiceBean.java:913)
	at edu.harvard.iq.dataverse.search.IndexServiceBean.indexDataset(IndexServiceBean.java:406)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:64)
	at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at sun.reflect.GeneratedMethodAccessor115.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
	at sun.reflect.GeneratedMethodAccessor120.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:88)
	at com.sun.proxy.$Proxy283.indexDataset(Unknown Source)
	at edu.harvard.iq.dataverse.search.__EJB31_Generated__IndexServiceBean__Intf____Bean__.indexDataset(Unknown Source)
	at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetVersionCommand.execute(UpdateDatasetVersionCommand.java:146)
	at edu.harvard.iq.dataverse.engine.command.impl.UpdateDatasetVersionCommand.execute(UpdateDatasetVersionCommand.java:19)
	at edu.harvard.iq.dataverse.EjbDataverseEngine.submit(EjbDataverseEngine.java:224)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.glassfish.ejb.security.application.EJBSecurityManager.runMethod(EJBSecurityManager.java:1081)
	at org.glassfish.ejb.security.application.EJBSecurityManager.invoke(EJBSecurityManager.java:1153)
	at com.sun.ejb.containers.BaseContainer.invokeBeanMethod(BaseContainer.java:4786)
	at com.sun.ejb.EjbInvocation.invokeBeanMethod(EjbInvocation.java:656)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at org.jboss.weld.ejb.AbstractEJBRequestScopeActivationInterceptor.aroundInvoke(AbstractEJBRequestScopeActivationInterceptor.java:73)
	at org.jboss.weld.ejb.SessionBeanInterceptor.aroundInvoke(SessionBeanInterceptor.java:52)
	at sun.reflect.GeneratedMethodAccessor115.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.EjbInvocation.proceed(EjbInvocation.java:608)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.doCall(SystemInterceptorProxy.java:163)
	at com.sun.ejb.containers.interceptors.SystemInterceptorProxy.aroundInvoke(SystemInterceptorProxy.java:140)
	at sun.reflect.GeneratedMethodAccessor120.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.sun.ejb.containers.interceptors.AroundInvokeInterceptor.intercept(InterceptorManager.java:883)
	at com.sun.ejb.containers.interceptors.AroundInvokeChainImpl.invokeNext(InterceptorManager.java:822)
	at com.sun.ejb.containers.interceptors.InterceptorManager.intercept(InterceptorManager.java:369)
	at com.sun.ejb.containers.BaseContainer.__intercept(BaseContainer.java:4758)
	at com.sun.ejb.containers.BaseContainer.intercept(BaseContainer.java:4746)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:212)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:88)
	at com.sun.proxy.$Proxy339.submit(Unknown Source)
	at edu.harvard.iq.dataverse.__EJB31_Generated__EjbDataverseEngine__Intf____Bean__.submit(Unknown Source)
	at edu.harvard.iq.dataverse.batch.jobs.importer.filesystem.FileRecordWriter.updateDatasetVersion(FileRecordWriter.java:194)
	at edu.harvard.iq.dataverse.batch.jobs.importer.filesystem.FileRecordWriter.writeItems(FileRecordWriter.java:169)
	at com.ibm.jbatch.container.artifact.proxy.ItemWriterProxy.writeItems(ItemWriterProxy.java:71)
	at com.ibm.jbatch.container.impl.ChunkStepControllerImpl.writeChunk(ChunkStepControllerImpl.java:470)
	at com.ibm.jbatch.container.impl.ChunkStepControllerImpl.invokeChunk(ChunkStepControllerImpl.java:587)
	at com.ibm.jbatch.container.impl.ChunkStepControllerImpl.invokeCoreStep(ChunkStepControllerImpl.java:684)
	at com.ibm.jbatch.container.impl.BaseStepControllerImpl.execute(BaseStepControllerImpl.java:144)
	at com.ibm.jbatch.container.impl.ExecutionTransitioner.doExecutionLoop(ExecutionTransitioner.java:112)
	at com.ibm.jbatch.container.impl.JobThreadRootControllerImpl.originateExecutionOnThread(JobThreadRootControllerImpl.java:110)
	at com.ibm.jbatch.container.util.BatchWorkUnit.run(BatchWorkUnit.java:80)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at org.glassfish.enterprise.concurrent.internal.ManagedFutureTask.run(ManagedFutureTask.java:141)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
	at org.glassfish.enterprise.concurrent.ManagedThreadFactoryImpl$ManagedThread.run(ManagedThreadFactoryImpl.java:250)]]

**Comment by qqmyers** [comment]
@kcondon - which banner? When I run, I see the publish button disappear but don't see a banner (Dataset Locked – Publish workflow in progress) unless/until I refresh - the log shows a 'response has already been committed error'. It looks like the lock is set in the db though. Or is it an issue with the pid register lock (assuming enough files) after the prepublish workflow runs, or?

For the index issue - probably good to have a separate issue, but what should happen?

**Comment by kcondon** [comment]
@qqmyers I will track down the specifics. I'm not sure which lock is thrown but on /develop the dataset lock banner appears immediately after clicking publish but it does not in your branch. I'll query the open locks on that dataset using the api and let you know what it says.

I will open a separate ticket on that index error. I'm not sure what it should do: the package file is essentially a folder containing other files. So we could either skip it or try to index the contents. I'll check in with the group.

**Comment by kcondon** [comment]
@qqmyers 
{"status":"OK","data":[{"lockType":"Workflow","date":"Tue Oct 23 17:58:11 EDT 2018","user":"dataverseAdmin"}
![Screen Shot 2018-10-23 at 5.58.26 PM.png](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/5798d98d18ccf10e00ce215e/125516c66e82c728ace21e0d46db978826878de9bce1b057e40fa0c5710b76236456f06b7191a37df117234f5d595ecc3c4a4356bcbbd671f3e57862804e0ee8856f5c75903d47afed8a4bec04ec208b4b8de7bddb1ed2f42f67b6d5db1063474f4dcbed5727a19e41c95df24be24e54ad153e65e1722431.png)

**Comment by kcondon** [comment]
related index ticket: https://github.com/IQSS/dataverse/issues/5228

**Comment by kcondon** [comment]
@qqmyers So the only issue left is the missing lock banner and grayed out buttons. I need to verify this but since the lock exists in the db, it may be a page refresh not happening as it did previously.

**Comment by qqmyers** [comment]
@kcondon - yes that's the only issue. I'm seeing slightly different things on my 4.9.4-based branch - it's only when there are enough files to go asynchronous that the banner doesn't refresh. The solution isn't obvious to me at this point... There isn't anything directly related to page refresh or message creation that has changed that I know of, just updates related to reading/writing db changes... I'll keep looking, but I have to intersperse other things. Any thoughts welcome...

**Comment by kcondon** [comment]
@qqmyers Well for me it is while publishing with an rsync package file. The behavior you are describing with asynch and enough files is a different, designed behavior to account for file doi registration delay with many files. The actual lock type in that case would be different, pidRegister versus workflow.

We have encountered this state not appearing immediately before in other areas. Let me see whether a page refresh makes it appear in my context or if something more is needed. I can ask the group if they have suggestions. 

Take your time on fixing it -I just did not want to hold it up from my end.
